### PR TITLE
NLL: Suggest `ref mut` and `&mut self`

### LINKED ISSUE #51244: Wrong diagnostic label text due to text replacement
The following snippet:
```rust
fn main() {
    let ref my_ref @ _ = 0;
    *my_ref = 0
}
```
Produces the following output:
```
error[E0594]: cannot assign to immutable borrowed content `*my_ref`
 --> src/main.rs:3:5
  |
2 |     let ref my_ref @ _ = 0;
  |         -------------- consider changing this to `ref mut my_ref mut @ _`
3 |     *my_ref = 0
  |     ^^^^^^^^^^^ cannot borrow as mutable
```
Note that the label suggests invalid syntax. I believe this happens due to the text replacement in the following code:
https://github.com/rust-lang/rust/blob/e38554cd80af2e99d626230ad3531b55ee3de65c/src/librustc_borrowck/borrowck/mod.rs#L1210-L1219

### PR DESCRIPTION
**Pull request by ashtneoi**
Fixes #51244. Supersedes #51249, I think.

Under the old lexical lifetimes, the compiler provided helpful suggestions about adding `mut` when you tried to mutate a variable bound as `&self` or (explicit) `ref`. NLL doesn't have those suggestions yet. This pull request adds them.

I didn't bother making the help text exactly the same as without NLL, but I can if that's important.

(Originally this was supposed to be part of #51612, but I got bogged down trying to fit everything in one PR.)

### CODE REVIEW

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -1837,17 +1839,43 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                    )),
+
+                    Some(ClearCrossCrate::Set(mir::BindingForm::Var(mir::VarBindingForm {
+                        binding_mode: ty::BindingMode::BindByReference(_),
+                        ..
+                    }))) => suggest_ref_mut(self.tcx, local_decl),
+
+                    Some(ClearCrossCrate::Clear) => bug!("saw cleared local state"),
+
+                    None => bug!(),
```
Comment (ashtneoi) [inline src/librustc_mir/borrow_check/mod.rs]: Not really sure what the most idiomatic way to do this is. I could add `.as_ref().unwrap()` on line 1844, but that felt kinda ugly.

File: src/librustc_mir/borrow_check/mod.rs [hunk 2]
```diff
@@ -1889,48 +1918,60 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+            local_decl: &mir::LocalDecl<'tcx>,
+            opt_ty_info: Option<Span>,
         ) -> (Span, String) {
-            // This implementation attempts to emulate AST-borrowck prioritization
-            // by trying (3.), then (2.) and finally falling back on (1.).
             let locations = mir.find_assignments(local);
             if locations.len() > 0 {
                 let assignment_rhs_span = mir.source_info(locations[0]).span;
                 let snippet = tcx.sess.codemap().span_to_snippet(assignment_rhs_span);
                 if let Ok(src) = snippet {
```
Comment (ashtneoi) [inline src/librustc_mir/borrow_check/mod.rs]: It's not clear to me how `span_to_snippet` could fail, but I left the fallthrough alone just in case.

Comment (ashtneoi) [reply src/librustc_mir/borrow_check/mod.rs]: (instead of unwrapping the result)

**Comment by ashtneoi** [comment]
cc @pnkfelix if you're interested

**Comment by csmoe** [comment]
@ashtneoi wow, you moved further than me https://github.com/rust-lang/rust/pull/51249 when addressing the issue about old suggestion https://github.com/rust-lang/rust/issues/51244 , it's well covered by your `suggest_ref_mut` method, would you mind fixing that here?

**Comment by ashtneoi** [comment]
@csmoe [Does this look right?](https://github.com/ashtneoi/rust/commits/suggest-ref-mut-csmoe) I rebased your branch onto mine and updated the tests.

**Comment by ashtneoi** [comment]
Or did you want me to factor out the `ref` -> `ref mut` code?

**Comment by csmoe** [comment]
@ashtneoi  yep, I wanna introduce  your `suggest_ref_mut`(or likely) there since the old and my replacement solution is untidy.

**Comment by ashtneoi** [comment]
@csmoe Ok, I think I did it. Also fixed that "use a mutable reference instead: `x`" help message.

**Comment by pnkfelix** [comment]
Weird, I had tried something similar to your changes here, but I get hitting a problem where it was making erroneous suggestions for mutations of references in arm guards.

In particular for the test [issue-27282-reborrow-ref-mut-in-guard.rs](https://github.com/rust-lang/rust/blob/8b48b247dc24a40ce499415370bfd2b6cefa3a1b/src/test/ui/issue-27282-reborrow-ref-mut-in-guard.rs#L26), *my* version of this code would try to suggest changing the `ref mut r`, but there is (by design) actually no way for the user to change `r` in any way that would make `*r` mutable within the guard.

So I had been spending time putting in some infrastructure to be able to filter this case.

But it seems like your PR here does not run into that problem.

I have not managed to figure out *how* you have managed to avoid it. heh.

**Comment by pnkfelix** [comment]
So, the approach I was planning to use wasn't going to have the `local_decl.is_user_variable.is_some()` guard; it was going to do something different which would have allowed it to suggest `ref mut` for bindings that are bound at deeper levels within a pattern.

In particular, if you look at my versions of tests like [rfc-2005-default-binding-mode/enum.nll.stderr](https://github.com/pnkfelix/rust/blob/nll-diagnostics-suggest-ref-mut/src/test/ui/rfc-2005-default-binding-mode/enum.nll.stderr) and [rfc-2005-default-binding-mode/explicit-mut.nll.stderr](https://github.com/pnkfelix/rust/blob/nll-diagnostics-suggest-ref-mut/src/test/ui/rfc-2005-default-binding-mode/explicit-mut.nll.stderr), I have managed to provide suggestions for bindings like `Some(n)`, even though those are not tagged with `is_user_variable`.

Having said that, there is something buggy in my own approach (which is why I have not put up a PR for it yet) and your approach is clearly fixing the bug in question and helping us bring NLL closer to parity with AST-borrowck in diagnostic quality.

So I am going to r+ your PR which will let me put my own branch on the back-burner.

**Comment by ashtneoi** [comment]
@pnkfelix As I understand it, the check for `src.starts_with("ref")` (plus whitespace) in `suggest_ref_mut` is what keeps implicit ref bindings like `Some(n)` from triggering the `ref mut` suggestion. I was able to get messages like this...

```none
+	LL |         Some(n) => {
+	   |              - help: consider changing this to be a mutable reference: `<make this mutable> n`
16	LL |             *n += 1; //~ ERROR cannot assign to immutable
```

...by just returning Some instead of None from `suggest_ref_mut` when the binding doesn't start with `ref`. Should I submit a PR for that (with better wording, of course)?

**Comment by pnkfelix** [comment]
Hmm I had thought, beyond the problem you describe, that there was also an issue involving `is_user_variable` (and I had thought there was some reason to try to leverage the `internal` variable for some purpose here...).

But its possible my memory is faulty.

**Pull request merged**